hie-plugin-api/Haskell/Ide/Engine/Cradle.hs

```diff
@@ -57,6 +58,64 @@ isStackCradle = (`elem` ["stack", "Cabal-Helper-Stack", "Cabal-Helper-Stack-None
   . BIOS.actionName
   . BIOS.cradleOptsProg
 
+  -- | Check if the given cradle is a cabal cradle.
+-- This might be used to determine the GHC version to use on the project.
+-- If it is a stack-cradle, we have to use `stack path --compiler-exe`
+-- otherwise we may ask `ghc` directly what version it is.
+isCabalCradle :: Cradle -> Bool
+isCabalCradle =
+  (`elem`
+    ["cabal"
+    , "Cabal-Helper-Cabal-V1"
+    , "Cabal-Helper-Cabal-V2"
+    , "Cabal-Helper-Cabal-V1-Dir"
+    , "Cabal-Helper-Cabal-V2-Dir"
+    , "Cabal-Helper-Cabal-None"
+    ]
+  )
+  . BIOS.actionName
+  . BIOS.cradleOptsProg
+
+
+getProjectGhcPath :: Cradle -> IO (Maybe FilePath)
+getProjectGhcPath crdl = do
+  isStackInstalled <- isJust <$> findExecutable "stack"
+  isCabalInstalled <- isJust <$> findExecutable "cabal"
+  ghcpath <- if isStackCradle crdl && isStackInstalled
+    then
+      catch (Just <$> tryCommand "stack path --compiler-exe") $ \(_ :: IOException) -> do
+        errorm "Command `stack path --compiler-exe` failed."
+        return Nothing
+    else if isCabalCradle crdl && isCabalInstalled then do
+      ghcCabalVersion <- catch (Just <$> tryCommand "cabal v2-exec -v0 ghc -- --numeric-version") $ \(_ ::IOException) -> do
+        errorm "Command `cabal v2-exec -v0 ghc -- --numeric-version` failed."
+        return Nothing
+      case ghcCabalVersion of
+        Just ghcNumericVersion -> do
+          let ghcVersion = "ghc-" ++ ghcNumericVersion
+          logm $ "Ghc Version to find: " ++ ghcVersion
+          findExecutable ghcVersion
+        Nothing -> return Nothing
+    else do
+      logm "Neither cabal nor stack project, look for ghc project."
+      findExecutable "ghc"
+  logm $ "Found ghc path: " ++ show ghcpath
+  return ghcpath
+
+tryCommand :: String -> IO String
+tryCommand cmd =
+  T.unpack . T.strip .T.pack <$> readCreateProcess (shell cmd) ""
```

Comment:
I will change this to something like `String -> IO (ExitCode, String, String)` for easier logs. Any objections?

Reply:
Sounds ok to me